### ISSUE TITLE
Socket IO Reconnect After Message Queue Failure

### ISSUE THREAD

**Issue by sjrumsby** [miguelgrinberg/Flask-SocketIO#830]
Hi!

I'm having some issues with Flask-SocketIO gracefully handling Rabbit MQ service interruptions. I've creating an example repo that should illustrate what I'm seeing: https://github.com/sjrumsby/flask-mq. What I am seeing is that after the message queue connection finishes reconnecting, SocketIO is unable to write data to the socket for ~1 minute.

When the rabbit MQ server is rebooted, I see the following logs:

```
/Users/sjrumsby/src/flask-mq/venv/lib/python3.6/site-packages/amqp/connection.py:312: AMQPDeprecationWarning: The .transport attribute on the connection was accessed before
the connection was established.  This is supported for now, but will
be deprecated in amqp 2.2.0.

Since amqp 2.0 you have to explicitly call Connection.connect()
before using the connection.

  W_FORCE_CONNECT.format(attr=attr)))
Connection error while reading from queue
Traceback (most recent call last):
  File "/Users/sjrumsby/src/flask-mq/venv/lib/python3.6/site-packages/socketio/kombu_manager.py", line 98, in _listen
    message = queue.get(block=True)
  File "/Users/sjrumsby/src/flask-mq/venv/lib/python3.6/site-packages/kombu/simple.py", line 61, in get
    self.channel.connection.client.drain_events(timeout=remaining)
  File "/Users/sjrumsby/src/flask-mq/venv/lib/python3.6/site-packages/kombu/connection.py", line 301, in drain_events
    return self.transport.drain_events(self.connection, **kwargs)
  File "/Users/sjrumsby/src/flask-mq/venv/lib/python3.6/site-packages/kombu/transport/pyamqp.py", line 103, in drain_events
    return connection.drain_events(**kwargs)
  File "/Users/sjrumsby/src/flask-mq/venv/lib/python3.6/site-packages/amqp/connection.py", line 491, in drain_events
    while not self.blocking_read(timeout):
  File "/Users/sjrumsby/src/flask-mq/venv/lib/python3.6/site-packages/amqp/connection.py", line 497, in blocking_read
    return self.on_inbound_frame(frame)
  File "/Users/sjrumsby/src/flask-mq/venv/lib/python3.6/site-packages/amqp/method_framing.py", line 55, in on_frame
    callback(channel, method_sig, buf, None)
  File "/Users/sjrumsby/src/flask-mq/venv/lib/python3.6/site-packages/amqp/connection.py", line 501, in on_inbound_method
    method_sig, payload, content,
  File "/Users/sjrumsby/src/flask-mq/venv/lib/python3.6/site-packages/amqp/abstract_channel.py", line 128, in dispatch_method
    listener(*args)
  File "/Users/sjrumsby/src/flask-mq/venv/lib/python3.6/site-packages/amqp/connection.py", line 623, in _on_close
    (class_id, method_id), ConnectionError)
amqp.exceptions.ConnectionForced: (0, 0): (320) CONNECTION_FORCED - broker forced connection closure with reason 'shutdown'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/sjrumsby/src/flask-mq/venv/lib/python3.6/site-packages/socketio/kombu_manager.py", line 100, in _listen
    yield message.payload
  File "/Users/sjrumsby/src/flask-mq/venv/lib/python3.6/site-packages/kombu/simple.py", line 24, in __exit__
    self.close()
  File "/Users/sjrumsby/src/flask-mq/venv/lib/python3.6/site-packages/kombu/simple.py", line 92, in close
    self.consumer.cancel()
  File "/Users/sjrumsby/src/flask-mq/venv/lib/python3.6/site-packages/kombu/messaging.py", line 488, in cancel
    cancel(tag)
  File "/Users/sjrumsby/src/flask-mq/venv/lib/python3.6/site-packages/amqp/channel.py", line 1438, in basic_cancel
    wait=None if nowait else spec.Basic.CancelOk,
  File "/Users/sjrumsby/src/flask-mq/venv/lib/python3.6/site-packages/amqp/abstract_channel.py", line 59, in send_method
    return self.wait(wait, returns_tuple=returns_tuple)
  File "/Users/sjrumsby/src/flask-mq/venv/lib/python3.6/site-packages/amqp/abstract_channel.py", line 79, in wait
    self.connection.drain_events(timeout=timeout)
  File "/Users/sjrumsby/src/flask-mq/venv/lib/python3.6/site-packages/amqp/connection.py", line 491, in drain_events
    while not self.blocking_read(timeout):
  File "/Users/sjrumsby/src/flask-mq/venv/lib/python3.6/site-packages/amqp/connection.py", line 496, in blocking_read
    frame = self.transport.read_frame()
  File "/Users/sjrumsby/src/flask-mq/venv/lib/python3.6/site-packages/amqp/transport.py", line 243, in read_frame
    frame_header = read(7, True)
  File "/Users/sjrumsby/src/flask-mq/venv/lib/python3.6/site-packages/amqp/transport.py", line 426, in _read
    raise IOError('Socket closed')
OSError: Socket closed
Sleeping 2.0s
Traceback (most recent call last):
  File "/Users/sjrumsby/src/flask-mq/venv/lib/python3.6/site-packages/kombu/utils/functional.py", line 332, in retry_over_time
    return fun(*args, **kwargs)
  File "/Users/sjrumsby/src/flask-mq/venv/lib/python3.6/site-packages/kombu/connection.py", line 261, in connect
    return self.connection
  File "/Users/sjrumsby/src/flask-mq/venv/lib/python3.6/site-packages/kombu/connection.py", line 802, in connection
    self._connection = self._establish_connection()
  File "/Users/sjrumsby/src/flask-mq/venv/lib/python3.6/site-packages/kombu/connection.py", line 757, in _establish_connection
    conn = self.transport.establish_connection()
  File "/Users/sjrumsby/src/flask-mq/venv/lib/python3.6/site-packages/kombu/transport/pyamqp.py", line 130, in establish_connection
    conn.connect()
  File "/Users/sjrumsby/src/flask-mq/venv/lib/python3.6/site-packages/amqp/connection.py", line 308, in connect
    self.drain_events(timeout=self.connect_timeout)
  File "/Users/sjrumsby/src/flask-mq/venv/lib/python3.6/site-packages/amqp/connection.py", line 491, in drain_events
    while not self.blocking_read(timeout):
  File "/Users/sjrumsby/src/flask-mq/venv/lib/python3.6/site-packages/amqp/connection.py", line 496, in blocking_read
    frame = self.transport.read_frame()
  File "/Users/sjrumsby/src/flask-mq/venv/lib/python3.6/site-packages/amqp/transport.py", line 243, in read_frame
    frame_header = read(7, True)
  File "/Users/sjrumsby/src/flask-mq/venv/lib/python3.6/site-packages/amqp/transport.py", line 426, in _read
    raise IOError('Socket closed')
OSError: Socket closed
Sleeping 4.0s
Traceback (most recent call last):
  File "/Users/sjrumsby/src/flask-mq/venv/lib/python3.6/site-packages/kombu/utils/functional.py", line 332, in retry_over_time
    return fun(*args, **kwargs)
  File "/Users/sjrumsby/src/flask-mq/venv/lib/python3.6/site-packages/kombu/connection.py", line 261, in connect
    return self.connection
  File "/Users/sjrumsby/src/flask-mq/venv/lib/python3.6/site-packages/kombu/connection.py", line 802, in connection
    self._connection = self._establish_connection()
  File "/Users/sjrumsby/src/flask-mq/venv/lib/python3.6/site-packages/kombu/connection.py", line 757, in _establish_connection
    conn = self.transport.establish_connection()
  File "/Users/sjrumsby/src/flask-mq/venv/lib/python3.6/site-packages/kombu/transport/pyamqp.py", line 130, in establish_connection
    conn.connect()
  File "/Users/sjrumsby/src/flask-mq/venv/lib/python3.6/site-packages/amqp/connection.py", line 308, in connect
    self.drain_events(timeout=self.connect_timeout)
  File "/Users/sjrumsby/src/flask-mq/venv/lib/python3.6/site-packages/amqp/connection.py", line 491, in drain_events
    while not self.blocking_read(timeout):
  File "/Users/sjrumsby/src/flask-mq/venv/lib/python3.6/site-packages/amqp/connection.py", line 496, in blocking_read
    frame = self.transport.read_frame()
  File "/Users/sjrumsby/src/flask-mq/venv/lib/python3.6/site-packages/amqp/transport.py", line 243, in read_frame
    frame_header = read(7, True)
  File "/Users/sjrumsby/src/flask-mq/venv/lib/python3.6/site-packages/amqp/transport.py", line 426, in _read
    raise IOError('Socket closed')
OSError: Socket closed
Sleeping 6.0s
Traceback (most recent call last):
  File "/Users/sjrumsby/src/flask-mq/venv/lib/python3.6/site-packages/kombu/utils/functional.py", line 332, in retry_over_time
    return fun(*args, **kwargs)
  File "/Users/sjrumsby/src/flask-mq/venv/lib/python3.6/site-packages/kombu/connection.py", line 261, in connect
    return self.connection
  File "/Users/sjrumsby/src/flask-mq/venv/lib/python3.6/site-packages/kombu/connection.py", line 802, in connection
    self._connection = self._establish_connection()
  File "/Users/sjrumsby/src/flask-mq/venv/lib/python3.6/site-packages/kombu/connection.py", line 757, in _establish_connection
    conn = self.transport.establish_connection()
  File "/Users/sjrumsby/src/flask-mq/venv/lib/python3.6/site-packages/kombu/transport/pyamqp.py", line 130, in establish_connection
    conn.connect()
  File "/Users/sjrumsby/src/flask-mq/venv/lib/python3.6/site-packages/amqp/connection.py", line 308, in connect
    self.drain_events(timeout=self.connect_timeout)
  File "/Users/sjrumsby/src/flask-mq/venv/lib/python3.6/site-packages/amqp/connection.py", line 491, in drain_events
    while not self.blocking_read(timeout):
  File "/Users/sjrumsby/src/flask-mq/venv/lib/python3.6/site-packages/amqp/connection.py", line 496, in blocking_read
    frame = self.transport.read_frame()
  File "/Users/sjrumsby/src/flask-mq/venv/lib/python3.6/site-packages/amqp/transport.py", line 243, in read_frame
    frame_header = read(7, True)
  File "/Users/sjrumsby/src/flask-mq/venv/lib/python3.6/site-packages/amqp/transport.py", line 426, in _read
    raise IOError('Socket closed')
OSError: Socket closed
```

This seems reasonable since the docker container takes ~10 seconds to reboot, so the back off trying to reconnect makes sense. However, once the connection is established again when I send data from the client browser back to Flask, Flask reports receiving them and attempting to write the data back but the packets are never sent. After some period of time (approximately 1 minute), another stack trace appears and then things go back to normal:

```
943a18bdbc994cb09bbaf3c2bc24fd52: Received packet MESSAGE data 2/test,["my_event",{"data":"jsdhfkdsfkjds"}]
received event "my_event" from 943a18bdbc994cb09bbaf3c2bc24fd52 [/test]
emitting event "my_response" to 943a18bdbc994cb09bbaf3c2bc24fd52 [/test]
943a18bdbc994cb09bbaf3c2bc24fd52: Received packet PING data None
943a18bdbc994cb09bbaf3c2bc24fd52: Sending packet PONG data None
943a18bdbc994cb09bbaf3c2bc24fd52: Received packet MESSAGE data 2/test,["my_event",{"data":"jsdhfkdsfkjds"}]
received event "my_event" from 943a18bdbc994cb09bbaf3c2bc24fd52 [/test]
emitting event "my_response" to 943a18bdbc994cb09bbaf3c2bc24fd52 [/test]
943a18bdbc994cb09bbaf3c2bc24fd52: Received packet MESSAGE data 2/test,["my_event",{"data":"jsdhfkdsfkjds"}]
received event "my_event" from 943a18bdbc994cb09bbaf3c2bc24fd52 [/test]
emitting event "my_response" to 943a18bdbc994cb09bbaf3c2bc24fd52 [/test]
943a18bdbc994cb09bbaf3c2bc24fd52: Received packet MESSAGE data 2/test,["my_event",{"data":"jsdhfkdsfkjds"}]
received event "my_event" from 943a18bdbc994cb09bbaf3c2bc24fd52 [/test]
emitting event "my_response" to 943a18bdbc994cb09bbaf3c2bc24fd52 [/test]
Sleeping 0s
Traceback (most recent call last):
  File "/Users/sjrumsby/src/flask-mq/venv/lib/python3.6/site-packages/kombu/connection.py", line 494, in _ensured
    return fun(*args, **kwargs)
  File "/Users/sjrumsby/src/flask-mq/venv/lib/python3.6/site-packages/kombu/messaging.py", line 181, in publish
    exchange_name, declare,
  File "/Users/sjrumsby/src/flask-mq/venv/lib/python3.6/site-packages/kombu/messaging.py", line 203, in _publish
    mandatory=mandatory, immediate=immediate,
  File "/Users/sjrumsby/src/flask-mq/venv/lib/python3.6/site-packages/amqp/channel.py", line 1732, in _basic_publish
    (0, exchange, routing_key, mandatory, immediate), msg
  File "/Users/sjrumsby/src/flask-mq/venv/lib/python3.6/site-packages/amqp/abstract_channel.py", line 50, in send_method
    conn.frame_writer(1, self.channel_id, sig, args, content)
  File "/Users/sjrumsby/src/flask-mq/venv/lib/python3.6/site-packages/amqp/method_framing.py", line 166, in write_frame
    write(view[:offset])
  File "/Users/sjrumsby/src/flask-mq/venv/lib/python3.6/site-packages/amqp/transport.py", line 275, in write
    self._write(s)
  File "/Users/sjrumsby/src/flask-mq/venv/lib/python3.6/site-packages/eventlet/greenio/base.py", line 401, in sendall
    tail = self.send(data, flags)
  File "/Users/sjrumsby/src/flask-mq/venv/lib/python3.6/site-packages/eventlet/greenio/base.py", line 395, in send
    return self._send_loop(self.fd.send, data, flags)
  File "/Users/sjrumsby/src/flask-mq/venv/lib/python3.6/site-packages/eventlet/greenio/base.py", line 382, in _send_loop
    return send_method(data, *args)
BrokenPipeError: [Errno 32] Broken pipe
943a18bdbc994cb09bbaf3c2bc24fd52: Sending packet MESSAGE data 2/test,["my_response",{"data":"jsdhfkdsfkjds","count":8}]
```

Here are the versions of packages I'm using:

```
$ pip list
DEPRECATION: The default format will switch to columns in the future. You can use --format=(legacy|columns) (or define a format=(legacy|columns) in your pip.conf under the [list] section) to disable this warning.
amqp (2.3.2)
Click (7.0)
dnspython (1.15.0)
eventlet (0.24.1)
Flask (1.0.2)
Flask-SocketIO (3.0.2)
greenlet (0.4.15)
gunicorn (19.9.0)
itsdangerous (1.1.0)
Jinja2 (2.10)
kombu (4.2.1)
MarkupSafe (1.1.0)
monotonic (1.5)
pip (9.0.1)
python-engineio (2.3.2)
python-socketio (2.0.0)
setuptools (28.8.0)
six (1.11.0)
vine (1.1.4)
Werkzeug (0.14.1)
```

Any help understanding what is going on here, or tracking down what the problem is would be greatly appreciated.

Thanks!!

**Comment by miguelgrinberg** [comment]
I'm not sure what this can be. The code is supposed to reconnect to the rabbit queue whenever a connection fails (code [here](https://github.com/miguelgrinberg/python-socketio/blob/master/socketio/kombu_manager.py#L87)). So unfortunately I cannot tell you what the problem is. It would seem as if the broken pipe error that you are getting isn't detected as an error requiring the connection to be re-established, which is weird. That would point to a problem in the Kombu package.

**Comment by sjrumsby** [comment]
Hi Miguel,

Thanks for getting back to us. We've since diagnoses the problem and have a solution, I'll include it here in case anyone comes across this issue in the future.

The provided example was a slightly simplified example from our actual production stack (we're using an MQ cluster in production). After some further debugging, we determined that the queues were not being mirrored across the cluster. In our simplified example, I'm guessing the queue would not persist through a docker stop and start as well.

For our purposes, we were able to set the `ha-mode` in RabbitMQ to `all` to mirror our queues throughout the cluster, which gives us the high-availability we need for our application. 

I think there is still an issue with the underlying kombu package (or perhaps RabbitMQ) in that it seems to think it's connected to the MQ server and trying to write messages, despite the queue it's writing to no longer existing on node.

Thanks again for the input and the great work on this package!

-Sean
